apps/user/serializers.py

```diff
@@ -225,3 +225,40 @@ def validate(self, attrs):
             user.save()
             return attrs
         raise serializers.ValidationError(ugettext('The token is invalid.'))
+
+
+class UpdateProfileSerializer(serializers.ModelSerializer):
+    institution = InstitutionUpdateSerializer(required=False)
+    school = SchoolUpdateSerializer(required=False)
+    publisher = PublisherUpdateSerializer(required=False)
+
+    class Meta:
+        model = User
+        fields = ('first_name', 'last_name', 'phone_number', 'image', 'publisher', 'school', 'institution')
+        extra_kwargs = {'password': {'write_only': True}}
+
+    def save(self, **kwargs):
+        instance = self.context['request'].user
+        institution_data = self.validated_data.pop('institution', None)
+        school_data = self.validated_data.pop('school', None)
+        publisher_data = self.validated_data.pop('publisher', None)
+        print(instance.user_type, User.UserType.INSTITUTIONAL_USER.value)
+        if institution_data and instance.user_type == User.UserType.INSTITUTIONAL_USER.value:
+            institution_data['municipality'] = institution_data['municipality'].id
+            serializer = InstitutionSerializer(data=institution_data, instance=instance.institution, partial=True)
+            if not serializer.is_valid():
+                raise serializers.ValidationError(serializer.errors)
+            serializer.save()
+        if school_data and instance.user_type == User.UserType.SCHOOL_ADMIN.value:
+            school_data['municipality'] = school_data['municipality'].id
+            serializer = SchoolSerializer(data=school_data, instance=instance.school, partial=True)
+            if not serializer.is_valid():
+                raise serializers.ValidationError(serializer.errors)
+            serializer.save()
+        if publisher_data and instance.user_type == User.UserType.PUBLISHER.value:
+            publisher_data['municipality'] = publisher_data['municipality'].id
+            serializer = InstitutionSerializer(data=publisher_data, instance=instance.publisher, partial=True)
+            if not serializer.is_valid():
+                raise serializers.ValidationError(serializer.errors)
+            serializer.save()
+        return super().update(instance, self.validated_data)
```

Comment:
Can we do something like this? That is not readable. Or use writable nested serializers?
```suggestion
    def _save_profile_data(self, user, data):
        data_key = {  # Used to retrieve data from validated_data
            User.UserType.INSTITUTIONAL_USER.value: 'institution',
            User.UserType.SCHOOL_ADMIN.value: 'school',
            User.UserType.PUBLISHER.value: 'publisher',
        }.get(instance.user_type)
        serializer_class = {  # Used to create/update profile data.
            User.UserType.INSTITUTIONAL_USER.value: InstitutionSerializer,
            User.UserType.SCHOOL_ADMIN.value: SchoolSerializer,
            User.UserType.PUBLISHER.value: InstitutionSerializer,  # FIXME: Is this correct?
        }.get(instance.user_type)
        current_instance_field = {  # Used to get current instance from user
            User.UserType.INSTITUTIONAL_USER.value: 'institution',
            User.UserType.SCHOOL_ADMIN.value: 'school',
            User.UserType.PUBLISHER.value: 'publisher',
        }.get(instance.user_type)

        instance = getattr(user, current_instance_field)  # eg: user.institution
        data = self.validated_data.pop(data_key, None)  # eg: institution data
        data['municipality'] = data['municipality'].id  # FIXME: WHY?
        serializer = serializer_class(data=data, instance=instance, partial=True)
        if not serializer.is_valid():
            raise serializers.ValidationError(serializer.errors)
        return serializer.save()

    def save(self, **kwargs):
        instance = self.context['request'].user
        self._save_profile_data(user, self.validated_data)
        return super().update(instance, self.validated_data)
```